### PR TITLE
I1357 - support for volumes

### PR DESCRIPTION
**Pull request by hillalex**
First review and merge https://github.com/vimc/bb8/pull/1

### CODE REVIEW

File: backup.py [hunk 1]
```diff
@@ -50,19 +53,43 @@ def run_rsync(settings, path):
...
+
+def run_rsync_from_container(settings, source_volume):
+    starport = settings.starport
+    docker_ssh_key_path = "/etc/id_rsa"
+
+    cmd = ["docker", "run", "--rm", "-i", "-t",
+           "-v", "{}:/{}".format(source_volume, source_volume),
+           "-v", "{}:{}".format(expanduser("~/.ssh/known_hosts"), "/root/.ssh/known_hosts"),
+           "-v", "{}:{}".format(abspath(settings.ssh_key_path), docker_ssh_key_path),
+           "instrumentisto/rsync-ssh"]
```
Comment (MartinEden) [inline backup.py]: How did you choose this image?

File: backup.py [hunk 2]
```diff
@@ -50,19 +53,43 @@ def run_rsync(settings, path):
...
+
+    cmd.extend(rsync_cmd(docker_ssh_key_path, source_volume, starport))
```
Comment (MartinEden) [inline backup.py]: Could this just be `cmd += rsync_cmd(docker_ssh_key_path, source_volume, starport)`

File: targets.py [hunk 3]
```diff
@@ -13,3 +15,25 @@ def before_restore(self):
...
+
+    def _volume_exists(self):
+        text = run(
+            ["docker", "volume", "ls", "-q"],
+            stdout=PIPE, universal_newlines=True
+        ).stdout
+        names = text.split('\n')
+        return self.name in names
+
+    def before_restore(self):
```
Comment (MartinEden) [inline targets.py]: Is this ever invoked?

**Review by MartinEden — CHANGES_REQUESTED** [review]

File: backup.py [hunk 4]
```diff
@@ -50,19 +53,43 @@ def run_rsync(settings, path):
...
+    starport = settings.starport
+    cmd = rsync_cmd(settings.ssh_key_path, path, starport)
+    run_cmd_with_logging(cmd)
+
+
+def run_rsync_from_container(settings, source_volume):
+    starport = settings.starport
+    docker_ssh_key_path = "/etc/id_rsa"
+
+    cmd = ["docker", "run", "--rm", "-i", "-t",
```
Comment (richfitz) [inline backup.py]: Is there any reason not to use the python docker module?

File: backup.py (hunk 1 above)
Comment (hillalex) [reply backup.py]: Pretty randomly, just from the top google hits for rsync ssh docker image! Its a very simple image, we could also create it ourselves.

File: backup.py (hunk 2 above)
Comment (hillalex) [reply backup.py]: Probably :)

File: targets.py (hunk 3 above)
Comment (hillalex) [reply targets.py]: No, think this was just copied over from previous backup script, but we can delete it for now until we come to write the restore part of bb8.

File: backup.py (hunk 4 above)
Comment (hillalex) [reply backup.py]: I don't know, is there an example of its usage in Montagu somewhere so I can compare?

Comment (richfitz) [reply backup.py]: service.py in montagu uses it quite a bit (https://github.com/vimc/montagu/blob/master/src/service.py)

The docs for the module are unusually readable https://docker-py.readthedocs.io/en/stable/ and I've found the api nicer than shelling out

**Review by MartinEden — APPROVED** [review]

**Pull request merged**